src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+    uint16_t proposedBlockSize = transferBlockSize;
+    TransferSession::TransferAcceptData acceptData;
+    acceptData.ControlMode    = respondingSender.GetControlMode();
+    acceptData.StartOffset    = proposedOffset;
+    acceptData.Length         = 0; // Indefinite length
+    acceptData.MaxBlockSize   = proposedBlockSize;
+    acceptData.Metadata       = reinterpret_cast<uint8_t *>(testMetadata);
+    acceptData.MetadataLength = 11;
+
+    err = respondingSender.AcceptTransfer(acceptData);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Verify Sender emits ReceiveAccept message for sending
+    respondingSender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle acceptMsgToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !acceptMsgToSend.IsNull());
+
+    // Pass Accept message to initiatingReceiver
+    err = initiatingReceiver.HandleMessageReceived(acceptMsgToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Verify received ReceiveAccept.
+    // Client may want to inspect TransferControl, MaxBlockSize, StartOffset, Length, and Metadata, and may choose to reject the
+    // Transfer at this point.
+    initiatingReceiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_AcceptReceived);
+    NL_TEST_ASSERT(inSuite, outEvent.transferAcceptData.ControlMode == kControl_ReceiverDrive);
+    NL_TEST_ASSERT(inSuite, outEvent.transferAcceptData.StartOffset == proposedOffset);
+    NL_TEST_ASSERT(inSuite, outEvent.transferAcceptData.Length == proposedLength);
+    NL_TEST_ASSERT(inSuite, outEvent.transferAcceptData.MaxBlockSize == transferBlockSize);
+    System::PacketBufferHandle rcvdAcceptMsgData = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !rcvdAcceptMsgData.IsNull());
+    NL_TEST_ASSERT(inSuite,
+                   !memcmp(testMetadata, outEvent.transferAcceptData.Metadata, outEvent.transferAcceptData.MetadataLength));
+
+    // Test BlockQuery -> Block -> BlockAck
+    SendAndVerifyQuery(inSuite, inContext, respondingSender, initiatingReceiver, outEvent);
+    SendAndVerifyArbitraryBlock(inSuite, inContext, respondingSender, initiatingReceiver, outEvent, false);
+    numBlocksSent++;
+    SendAndVerifyBlockAck(inSuite, inContext, respondingSender, initiatingReceiver, outEvent);
+
+    // Test multiple Blocks sent and received (last Block is BlockEOF)
+    while (numBlocksSent < numBlockSends)
+    {
+        bool isEof = (numBlocksSent == numBlockSends - 1);
+
+        SendAndVerifyQuery(inSuite, inContext, respondingSender, initiatingReceiver, outEvent);
+        SendAndVerifyArbitraryBlock(inSuite, inContext, respondingSender, initiatingReceiver, outEvent, isEof);
+
+        numBlocksSent++;
+    }
+
+    // Verify last block was BlockEOF, then verify that PrepareBlockAckEOF() emits BlockAckEOF message
```

Comment:
The PrepareBlockAck should magically send a BlockEOF if the last block received had been a BlockEOF.